### PR TITLE
[TECHDEBT] Consolidated PostgresContext and PostgresDB (#149)

### LINKED ISSUE #149: [Persistence] Fix Persistence Module TODOs - Consolidate Postgres
# Objective
A general cleanup issue is needed to tackle TODO's and ensure the persistence module is usable/readable by consolidating `PostgresContext` and `PostgresDb` as part of https://github.com/pokt-network/pocket/issues/172. 

# Origin Document
Should follow issue-#128, issue-#105, issue-#147 and issue-#148 the persistence module is messier and more difficult to understand than the developers would want for organic external contribution.

# Goals / Deliverables
- [ ] Run `log.
- [ ] Finish outstanding TODO's to ensure end-to-end functionality
- [ ] Consolidate `PostgresContext` and `PostgresDb` like so:
```go
type PostgresContext struct {
	height int64
	tx         pgx.Tx
	blockstore kvstore.KVStore	
}
```
- [ ] Avoid exposing `PostgresContext` and/or `PostgresDb` for testing purposes

## General issue checklist
- [ ] Update the appropriate CHANGELOG
- [ ] Update the README
- [ ] If applicable, update the source code tree explanation
- [ ] If applicable, add or update a state, sequence or flowchart diagram using [mermaid](https://mermaid-js.github.io/mermaid/)
- [ ] Update any relevant global documentation & references
- [ ] Document small issues / TODOs along the way
---

Creator: @andrewnguyen22
Co-creator: @Olshansk

### PR DESCRIPTION
**Pull request by Olshansk**
## Description

Tend to tech debt in the `persistence` module related to simplifying the `PostgresContext` struct.

## Issue

Fixes #149

## Type of change

Please mark the relevant option(s):

- [ ] New feature, functionality or library
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Code health or cleanup
- [ ] Major breaking change
- [ ] Documentation
- [ ] Other: <REPLACE_ME>

## List of changes

- Consolidates `PostgresContext` and `PostgresDB`
- Reduces scope of `Tx` and `BlockStore`
- Improve error handling in `NewFuzzTestPostgresContext`

## Testing

- [x] `make test_all`
- [ ] [LocalNet](https://github.com/pokt-network/pocket/blob/main/docs/development/README.md) w/ all of the steps outlined in the `README`

## Checklist

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have tested my changes using the available tooling
- [ ] If applicable, I have made corresponding changes to related local or global README
- [ ] If applicable, I have added new diagrams using [mermaid.js](https://mermaid-js.github.io)
- [ ] If applicable, I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: shared/node.go [hunk 1]
```diff
@@ -96,7 +96,7 @@ func (node *Node) Start() error {
 	signalNodeStartedEvent := &debug.PocketEvent{Topic: debug.PocketTopic_POCKET_NODE_TOPIC, Data: nil}
 	node.GetBus().PublishEventToBus(signalNodeStartedEvent)
 
-	log.Println("About to start pocket node main loop...")
+	log.Println("About to start pocket node main loop..")
```
Comment (andrewnguyen22) [inline shared/node.go:99]: Is 2 dots more semantic than 3?

File: persistence/test/setup_test.go [hunk 2]
```diff
@@ -65,32 +65,37 @@ func NewTestPostgresContext(t *testing.T, height int64) *persistence.PostgresCon
...
+		fmt.Printf("Error creating new context: %v\n", err)
+		f.FailNow()
 	}
-	db := persistence.PostgresContext{
-		Height: height,
-		DB:     ctx.(persistence.PostgresContext).DB,
+
+	db, ok := ctx.(persistence.PostgresContext)
+	if !ok {
+		fmt.Println("Error casting RW context to Postgres context")
```
Comment (andrewnguyen22) [inline persistence/test/setup_test.go:88]: Needs to follow a pattern.

`log.Fatal()` same for all `fmt.Print...`

File: persistence/CHANGELOG.md [hunk 3]
```diff
@@ -7,12 +7,18 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [Unreleased]
 
+## [0.0.0.6] - 2022-09-14
```
Comment (andrewnguyen22) [inline persistence/CHANGELOG.md:10]: Skipped version 0.0.0.5?

File: persistence/context.go [hunk 4]
```diff
@@ -12,7 +12,7 @@ func (p PostgresContext) NewSavePoint(bytes []byte) error {
 
 func (p PostgresContext) RollbackToSavePoint(bytes []byte) error {
 	log.Println("TODO: RollbackToSavePoint not fully implemented")
-	return p.DB.Tx.Rollback(context.TODO())
+	return p.tx.Rollback(context.TODO())
```
Comment (andrewnguyen22) [inline persistence/context.go:15]: Choose/consolidate `tx` or `txn`, not both - to name this field

And use `p.GetTxn()` instead of accessing directly

**Review by andrewnguyen22 — CHANGES_REQUESTED** [review]

File: shared/node.go (hunk 1 above)
Comment (Olshansk) [reply shared/node.go:99]: Nope, was an accident.

Reverted.

File: persistence/test/setup_test.go (hunk 2 above)
Comment (Olshansk) [reply persistence/test/setup_test.go:88]: Done.

File: persistence/CHANGELOG.md (hunk 3 above)
Comment (Olshansk) [reply persistence/CHANGELOG.md:10]: Off by one error

File: persistence/context.go (hunk 4 above)
Comment (Olshansk) [reply persistence/context.go:15]: Done. I removed the error so let me know if that's kosher in your opinion or not.

**Review by andrewnguyen22 — CHANGES_REQUESTED** [review]
You didn't complete the original ask about consolidating `txn` and `tx`.

Let me add more detail:

```go
type PostgresContext struct {
	Height     int64 // TODO(olshansky): `Height` is only externalized for testing purposes. Replace with helpers...
	conn       *pgx.Conn
	tx         pgx.Tx                          // This field is `tx`
	blockstore kvstore.KVStore
}
```

```
// These functions use Txn() 

func (pg *PostgresContext) GetCtxAndTxn() (context.Context, pgx.Tx, error) {
	return context.TODO(), pg.GetTxn(), nil
}

func (pg *PostgresContext) GetTxn() pgx.Tx {
	return pg.tx
}
```

Let's consolidate this terminology in order to lower cognitive load and increase readability

Also, please ensure that the structure field uses the `GetTxn()` function instead of accessing the private field directly:

Ex.

```
if err := p.tx.Rollback(ctx); err != nil {
		return err
	}
```

**Comment by Olshansk** [comment]
Got it now, ty! I def like this better.

Also, pro tip for markdown. You can add syntax highlighting by specifying the language after the triple backtick:
Screen Shot 2022-09-19 at 3 22 09 PM
Screen Shot 2022-09-19 at 3 30 28 PM
Screen Shot 2022-09-19 at 3 22 18 PM

File: persistence/test/setup_test.go [hunk 5]
```diff
@@ -204,10 +206,11 @@ func fuzzSingleProtocolActor(
 			require.NoError(t, err)
 
 			require.ElementsMatch(t, newActor.Chains, newChains, "staked chains not updated")
+			require.NotContains(t, newActor.StakedTokens, "invalid")
 			// TODO(andrew): Use `require.Contains` instead
-			if strings.Contains(newActor.StakedTokens, "invalid") {
-				fmt.Println("")
-			}
+			// if strings.Contains(newActor.StakedTokens, "invalid") {
```
Comment (andrewnguyen22) [inline persistence/test/setup_test.go:211]: Why is this commented out?

Also, if you're going to touch this code, it seems like you could easily satisfy that TODO there.

Perhaps it's out of scope?

File: persistence/account.go [hunk 6]
```diff
@@ -23,7 +23,7 @@ func (p PostgresContext) GetAccountAmount(address []byte, height int64) (amount
 }
 
 func (p PostgresContext) getAccountAmountStr(address string, height int64) (amount string, err error) {
-	ctx, txn, err := p.DB.GetCtxAndTxn()
+	ctx, txn, err := p.GetCtxAndTx()
```
Comment (andrewnguyen22) [inline persistence/account.go:26]: I don't want to see txn anywhere in this code anymore. Let's keep it consistent

Examples:

```go
ctx, txn, err := p.GetCtxAndTx()
```

```go 
func (p *PostgresContext) GetChainsForActor(
	ctx context.Context,
	txn pgx.Tx,
```

**Review by andrewnguyen22 — CHANGES_REQUESTED** [review]
Please change this PR description to fill out the template provided

I don't believe it's as clear to the outsider.

Also, consider adding more detail to the changelog for tracking purposes

**Comment by andrewnguyen22** [comment]
Resolve conflicts then approve incoming!

**Review by andrewnguyen22 — CHANGES_REQUESTED** [review]
Resolve conflicts then approve incoming!

**Comment by Olshansk** [comment]
@andrewnguyen22 Merged with main and pushed.

**Review by andrewnguyen22 — APPROVED** [review]

**Pull request merged**